File: 2559-maximum-number-of-non-overlapping-palindrome-substrings/2559-maximum-number-of-non-overlapping-palindrome-substrings.py
class Solution:
    def maxPalindromes(self, s: str, k: int) -> int:
        res = 0
        lastp = -1

        def helper(l, r):
            nonlocal lastp, res
            while l >= 0 and r < len(s) and s[l] == s[r] and l > lastp:
                if r-l+1 >= k:
                    res += 1
                    lastp = r
                    break # find the shortest palindrome
                else:
                    l -= 1
                    r += 1
        
        for i in range(len(s)):
            helper(i, i)  # odd length
            helper(i, i+1)  # even length
        return res          

File: 2559-maximum-number-of-non-overlapping-palindrome-substrings/2559-maximum-number-of-non-overlapping-palindrome-substrings.go


func maxPalindromes(s string, k int) int {
	res := 0
	lastp := -1
	n := len(s)

	helper := func(l, r *int) {
		for *l >= 0 && *r < n && s[*l] == s[*r] && *l > lastp {
			if *r-*l+1 >= k {
				res++
				lastp = *r
				break // find the shortest palindrome
			} else {
				*l--
				*r++
			}
		}
	}

	for i := 0; i < len(s); i++ {
		l, r := i, i // odd length
		helper(&l, &r)
		l, r = i, i+1 // even length
		helper(&l, &r)
	}
	return res
}

